cmd/kubeadm/app/util/error.go

```diff
@@ -64,16 +66,35 @@ type preflightError interface {
 // checkErr formats a given error as a string and calls the passed handleErr
 // func with that string and an exit code.
 func checkErr(err error, handleErr func(string, int)) {
+
+	var msg string
+	if err != nil {
+		msg = err.Error()
+		// check if the verbosity level in klog is high enough and print a stack trace.
+		f := flag.CommandLine.Lookup("v")
+		if f != nil {
+			// assume that the "v" flag contains a parseable Int32 as per klog's "Level" type alias,
+			// thus no error from ParseInt is handled here.
+			if v, e := strconv.ParseInt(f.Value.String(), 10, 32); e == nil {
+				// https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md
+				// klog.V(5) - Trace level verbosity
+				if v > 4 {
```

Comment:
I'm still convinced that this would be a problem for behaviours that aren't easily reproducible.

cc @ereslibre @timothysc

Reply:
I'm also not sure about this altogether. I think in light of a core dump we want the user to report it as soon as happened instead of asking for a reproducer, that's why I think we should provide the core dump right when happened, regardless of what was the verbosity the command was run with.

Reply:
are you suggesting dumping a stack trace regardless of the `v` level?
i'm fine with that.

but please read the original issue:
https://github.com/kubernetes/kubeadm/issues/1266#issuecomment-442911995

> we should not forget about the users! A stack trace is too extreme to show the users and is really only necessary during an unexpected error. If the error is expected then we have the ability to handle it and print a user friendly error message. If it's not expected we can print a stack trace along with a link to kubeadm/issues and instructions to fill out an issue.

i personally think this is difficult and that all errors are unexpected if we bubble them all the way up the stack and the program exists with `exit(1)`.

Reply:
> I'm also not sure about this altogether. I think in light of a core dump we want the user to report it as soon as happened instead of asking for a reproducer, that's why I think we should provide the core dump right when happened, regardless of what was the verbosity the command was run with.

Then we would have a lot of users claiming, that kubeadm is always crashing even on simplest of errors. This is terrible image for a tool, that is targeting user friendliness and easiness of use.

> i personally think this is difficult and that all errors are unexpected if we bubble them all the way up the stack and the program exists with exit(1).

In general it's not a good practice in software engineering to just bail out with an error deep inside code. Cleanup operations are hampered, people can no longer use kubeadm as a library, etc.
In fact, using `CheckErr` inside Cobra `Run` commands should be avoided (possibly even banned). Ideally `CheckErr` should be called from `main` and `Run` should be substituted for `RunE`.

Reply:
We can always enable this, regardless of verbosity level, for all non-stable builds (CI, dirty git tree, alpha, possibly beta).

Reply:
> Ideally CheckErr should be called from main and Run should be substituted for RunE.

+1, can be tracked in a separate ticket.

Reply:
> We can always enable this, regardless of verbosity level, for all non-stable builds (CI, dirty git tree, alpha, possibly beta).

i don't think it's that easy with the current way artifacts are built.
perhaps if we start managing the kubeadm artifact our self once kubeadm moves out of tree.
but this part is unclear to me at this point.

Reply:
> Then we would have a lot of users claiming, that kubeadm is always crashing even on simplest of errors. This is terrible image for a tool, that is targeting user friendliness and easiness of use.

What I mean is that we could produce a stacktrace on every unexpected error in disk. Kubeadm's output could look like: `Something unexpected happen, please report this error. Please attach the file /tmp/------.trace to the bug report to provide further information`.

I have the impression that we are not talking about the same kind of errors:

1) Regular error, bubbled up, error reported to the user, no stacktrace. The error was handled without any issues.
2) Unexpected error, bubbled up, error reported to the user, stacktrace in tempfile. The error was not expected (e.g. assertion failed)
3) SIGSEGV (e.g. we dereferenced nil), unchecked bad cast... error reported to the user, stacktrace in tempfile.

1) will never write a stacktrace since it's a controlled error that was handled as expected. 2) and 3) is what would write the stacktrace somewhere the user could fetch and attach it to us in a bug report. @timstclair raised concerns about the safety of the contents, as long as it's not a real core dump but a plain stack trace that helps us understand what happened I think it should be fine, but that's just my 2 cents :)

Reply:
how do we distinguish between 1 and 2?

> SIGSEGV (e.g. we dereferenced nil), unchecked bad cast... error reported to the user, stacktrace in tempfile.

does that mean recovering from a go panic?

Reply:
> how do we distinguish between 1 and 2?

Perhaps we can think about making most of our errors of type 1) and catch 2) and 3) with `recover` at top-level, dumping the stack trace, maybe using https://golang.org/ref/spec#Handling_panics, didn't test what kind of errors this method catches though.

> does that mean recovering from a go panic?

Yes, if possible (unchecked).

Reply:
I was thinking in something along the lines of https://play.golang.org/p/tMpb_-6FtBV

Reply:
since there is no way to distinguish between 1 and 2, unless we start marking them as such (which we probably should not do) this means that a stack trace will only be generated for 3, which already happens today, but on the user screen. users today always provide panic stack traces for us and if we move to write to a file they might as well forget to do that (i'm pretty positive this can happen).

also, errors of types 3 i would qualify as unhandled and arguably kubeadm should never exit with unhandled errors.

the original ticket however attempts to propose dumping a stack trace for handled errors for two groups - developer vs regular user. V levels does exactly that.

so IMO, the two possible ways to move forward with this PR are:
- always dump handled error stack traces independent on V level
- dump handled error stack traces for high V level.

and write a separate ticket to fix the usage of `CheckErr()` across the board.

Reply:
I think I'm missing something here because I don't see a reason why would we want to dump a stack trace of a handled error even with a high verbosity level.

Please, don't consider my comments blocking in any way, I fear I'm not understanding the problem here.

Reply:
> the original ticket however attempts to propose dumping a stack trace for handled errors for two groups - developer vs regular user. V levels do exactly that.

to clarify, the original issue talks about expected vs unexpected errors, which i think are not that easy to distinguish.

https://github.com/kubernetes/kubeadm/issues/1266#issuecomment-442869697
https://github.com/kubernetes/kubeadm/issues/1266#issuecomment-442911995

but i think, all errors are unexpected, and the only expected outcome is a kubeadm command success.

so from my POV, it's a question whether we should always dump a stack trace for all errors or only dump it for high V levels which are meant for developers.

Reply:
TL;DR, but from the size of the thread here, it's probably best to discuss this in the office hours this week so we can reach some conclusion.

Reply:
added to agenda for this week